src/ILOManager.sol

```diff
@@ -49,206 +56,296 @@ contract ILOManager is IILOManager, Ownable, Initializable {
+        Project memory _project = _projects[projectId];
+        require(_project.status == ProjectStatus.INITIALIZED, 'NA');
+        uint160 sqrtPriceX96 = _project.initialPoolPriceX96;
+        PoolAddress.PoolKey memory poolKey = PoolAddress.PoolKey(
+            token,
+            _project.pairToken,
+            _project.fee
+        );
 
-        uint160 sqrtPriceX96 = _projects[projectId].initialPoolPriceX96;
-        PoolAddress.PoolKey memory poolKey = PoolAddress.PoolKey(saleToken, _projects[projectId].raiseToken, _projects[projectId].fee);
-        
         // flip price and tokens
-        if (saleToken > _projects[projectId].raiseToken) {
+        if (token > _project.pairToken) {
             (poolKey.token0, poolKey.token1) = (poolKey.token1, poolKey.token0);
-            sqrtPriceX96 = uint160(2**192 / sqrtPriceX96);
+            sqrtPriceX96 = uint160(2 ** 192 / sqrtPriceX96);
         }
 
-        address uniV3PoolAddress = _initUniV3PoolIfNecessary(poolKey, sqrtPriceX96);
+        address uniV3PoolAddress = _initUniV3PoolIfNecessary(
+            poolKey,
+            sqrtPriceX96
+        );
 
-        address[] memory initializedPools = _initializedILOPools[projectId];
-        require(initializedPools.length > 0, "NP");
-        for (uint256 i = 0; i < initializedPools.length; i++) {
-            IILOPool(initializedPools[i]).launch(projectId, uniV3PoolAddress, poolKey);
+        EnumerableSet.AddressSet
+            storage initializedPools = _initializedILOPools[projectId];
+        uint256 length = EnumerableSet.length(initializedPools);
+        require(length > 0, 'NP');
+        for (uint256 i = 0; i < length; i++) {
+            IILOPoolBase(EnumerableSet.at(initializedPools, i)).launch(
+                uniV3PoolAddress,
+                poolKey,
+                sqrtPriceX96
+            );
         }
         IUniswapV3Pool(uniV3PoolAddress).increaseObservationCardinalityNext(10);
-        emit ProjectLaunch(projectId, uniV3PoolAddress, saleToken);
+        emit ProjectLaunch(projectId, uniV3PoolAddress, token);
     }
-    
-    /// @inheritdoc IILOManager
-    function initProjectFee() external override view returns (uint256) {
-        return _initProjectFee;
+
+    function cancelProject(
+        string calldata projectId
+    ) external override onlyOwner {
+        Project storage _project = _projects[projectId];
+        require(_project.status == ProjectStatus.INITIALIZED, 'NA');
+        _project.status = ProjectStatus.CANCELLED;
+        emit ProjectCancelled(projectId);
+    }
+
+    function cancelPoolSale(
+        string calldata projectId,
+        address pool
+    ) external override onlyProjectAdmin(projectId) {
+        EnumerableSet.AddressSet
+            storage initializedPools = _initializedILOPools[projectId];
+        require(EnumerableSet.contains(initializedPools, pool), 'NP');
+        EnumerableSet.remove(initializedPools, pool);
+        IILOPoolSale(pool).cancel();
```

Comment:
for any reason?

Reply:
Cancel call other contract while remove is just update internal Set, Btw, 2 functions don't depend on other so they will be success or reverted anyway. 
Just for clearer logic flow: cancel pool -> remove pool in manager and some gas saved just in case cancel failed